lib/jobs/run-graph.js

```diff
@@ -43,57 +45,97 @@ function runGraphJobFactory(
+        RunGraphJob.super_.call(this, _logger || logger, options, context, taskId);
 
         if (!this.options.instanceId) {
             this.options.instanceId = uuid.v4();
         }
+
         this.graphId = this.options.instanceId;
-        this.options.graphOptions.instanceId = this.graphId;
-        this.graphTarget = this.options.graphOptions.target;
-        this.domain = this.options.domain || Constants.Task.DefaultDomain;
         this.parentGraphId = this.context.graphId;
         this.proxy = context.proxy;
+        this.graphPollInterval = this.options.graphPollInterval || 10000;
+
+        if(this.constructor.name === 'RunGraphJob') {
+            //only do these things for this class and not its subclasses
+            assert.string(this.options.graphName);
+            assert.object(this.options.graphOptions);
+            this.options.graphOptions.instanceId = this.graphId;
+            this.graphTarget = this.options.graphOptions.target;
+            this.domain = this.options.domain || Constants.Task.DefaultDomain;
+        }
     }
     util.inherits(RunGraphJob, BaseJob);
 
+
     /**
      * @memberOf RunGraphJob
      */
     RunGraphJob.prototype._run = function() {
         var self = this;
-
-        this._subscribeGraphFinished(function(status) {
-            if (status === Constants.Task.States.Succeeded) {
-                self._done();
-            } else {
-                self._done(new Error("Graph " + self.graphId + " failed with status " + status));
+        self.subscribeOwnGraph().then(function(ownGraph) {
+            self.subgraphPoll = self.setSubgraphPoll();
+            if (!ownGraph) {
+                return workflowTool.runGraph(
+                    self.graphTarget,
+                    self.options.graphName,
+                    self.options.graphOptions,
+                    self.domain,
+                    self.proxy,
+                    self.parentGraphId,
+                    self.taskId
+                );
             }
+        })
+        .catch(function(error) {
+           logger.error("Error starting graph for task.", {
+                    taskId: self.taskId,
+                    graphName: self.options.graphName,
+                    error: error
+                });
+            self._done(error);
         });
+    };
 
-        workflowTool.runGraph(
-            self.graphTarget,
-            self.options.graphName,
-            self.options.graphOptions,
-            self.domain,
-            self.proxy,
-            self.parentGraphId,
-            self.taskId
-        ).catch(function(error) {
-            logger.error("Error starting graph for task.", {
-                taskId: self.taskId,
-                graphName: self.options.graphName,
-                error: error
-            });
-            self._done(error);
+    RunGraphJob.prototype.finishWithState = function(subGraphStatus) {
+        if (subGraphStatus === Constants.Task.States.Succeeded) {
```

Comment:
@VulpesArtificem If `self._done` is called, shouldn't the interval timer also be cancelled?  It seems like there might be a small window for a race condition here where the timer expires between handling a graph finished event and calling the `_cleanup` method.  I think this would result in calling `_done` twice.